version to check

ksef opened the pull request:
Description     
The companies are all set and ready to rent out their cars. Until now, there was no opportunity to interact with potential customers in any way. Let's create a log-in option for the customers so that they can rent a car through the platform.

Objectives      
Create a table named CUSTOMER with the following columns:

ID column should be PRIMARY KEY and AUTO_INCREMENT with the type INT.
NAME column should be UNIQUE and NOT NULL with the type VARCHAR.
RENTED_CAR_ID should have the type INT. This column should be a FOREIGN KEY referring to the ID column of the CAR table, and this column can be NULL in case the customer didn't rent a car.
Update the main menu adding a couple of new options:

1. Log in as a manager
2. Log in as a customer
3. Create a customer
4. Exit

Log in as a manager should work the same as described in the previous stage.

If the option Create a customer was chosen, the program should prompt the user to enter a customer name and then save it in the database. By default, the column RENTED_CAR_ID of the customer should be NULL.

If the Log in as a customer option was chosen, the program should print the list of existing customers and prompt the user to choose one:

Choose a customer:
1. First customer
2. Second customer
3. Back
   
If the Back option was chosen, go back and print the main menu. If the customer list is empty, print the message The customer list is empty! and return to the main menu. If the user chooses one of the customers, print the customer menu:

7. Rent a car
8. Return a rented car
9. My rented car
10. Back
    
If the Back option was chosen, go back and print the main menu.

If the user chose My rented car, print the name of the car they rented and the company it belongs to:

Your rented car:        
Hyundai Venue       
Company:        
Car To Go       

In case a customer doesn't currently have any rented cars, print You didn't rent a car!.

If a customer wants to Return a rented car, the program should set the column RENTED_CAR_ID of the current customer to NULL. If the user tries to return a car that they didn't actually rent, print You didn't rent a car!.

If the customer wants to Rent a car, the program should print the list of available companies and prompt the user to choose one. If the company list is empty, the program should print the message The company list is empty! and return to the customer menu. If the customer has already rented a car, print You've already rented a car!.

If the customer chose a company that doesn't have any available cars at the moment, print No available cars in the 'Company name' company.

If the customer was lucky and the chosen company has cars to rent out, the program should print the cars ordered by their ID column and prompt the user to pick one. 

Remember that indexes in the list should start from 1:

Choose a car:
1. Hyundai Venue
2. Maruti Suzuki Dzire
3. Back

After the user has chosen a car, print the message You rented 'Car name' and then go to the customer menu. If the customer didn't like any of the options and wants to go Back, the program should take them back to the customer menu.

Note that a list numeration should always start with 1.

To pass the tests, you have to enable the auto-commit mode so that all changes are automatically saved in the database file. To do that, call the method connection.setAutoCommit(true) of the Connection object.
Examples
The greater-than symbol followed by a space > represents the user input. Note that it's not part of the input.

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 2

The customer list is empty!

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 3

Enter the customer name:
> First customer
The customer was added!

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 3

Enter the customer name:
> Second customer
The customer was added!

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 2

Customer list:
1. First customer
2. Second customer
0. Back
> 1

1. Rent a car
2. Return a rented car
3. My rented car
0. Back
> 3

You didn't rent a car!

1. Rent a car
2. Return a rented car
3. My rented car
0. Back
> 2

You didn't rent a car!

1. Rent a car
2. Return a rented car
3. My rented car
0. Back
> 0

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 0

ksef committed:
version to check

ksef committed:
refactored CarSharingEngine and DBManager

taras-nishchymenko commented on build.gradle:
```diff
@@ -5,15 +5,32 @@ plugins {
 dependencies {
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.7.2'
-    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.7.2'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.2'
```
Please group your implementation and test specific dependencies

ksef replied:
grouped

taras-nishchymenko commented on src/main/java/carsharing/CarSharingEngine.java:
```diff
@@ -0,0 +1,90 @@
+
+public class CarSharingEngine {
+
+    public static final String BACK = "0. Back";
```
Looks weird. Do we need it?

ksef replied:
deleted

taras-nishchymenko commented on src/main/java/carsharing/CarSharingEngine.java:
```diff
@@ -0,0 +1,90 @@
+    public static final String BACK = "0. Back";
+    private final Scanner scanner;
+    private final CustomerDAO customerDao;
+    private String pathToDb;
```
what is this?

ksef replied:
this is db name, renamed

taras-nishchymenko commented on src/main/java/carsharing/CarSharingEngine.java:
```diff
@@ -0,0 +1,90 @@
+    private final Scanner scanner;
+    private final CustomerDAO customerDao;
+    private String pathToDb;
+    private final DBManager DBManager;
```
why do we need it?

ksef replied:
It moved this to constructor. 
I need it to pass the dbName to DBManager

taras-nishchymenko commented on src/main/java/carsharing/configuration/DBManager.java:
```diff
@@ -0,0 +1,60 @@
+            REFERENCES car(id));
+            """;
+
+    protected static String DB_URL;
```
Why do we need to change state of static field?

ksef replied:
DB_URL = "jdbc:h2:./src/main/java/carsharing/db/" + name; 
We can change db name

taras-nishchymenko commented on src/main/java/carsharing/configuration/DBManager.java:
```diff
@@ -0,0 +1,60 @@
+import java.sql.Statement;
+
+@NoArgsConstructor
+public class DBManager {
```
Why our DAOs extend this class?

ksef replied:
I fixed this

taras-nishchymenko commented on src/main/java/carsharing/customer/CustomerDAOImpl.java:
```diff
@@ -0,0 +1,214 @@
+    private static final String INSERT_INTO_CUSTOMER_BY_NAME = "INSERT INTO customer (name) VALUES ?";
+
+    private final DBManager DBManager;
+    private Set<Integer> rentedCars;
```
Why do we need store state outside of DB?

ksef replied:
This is car id, we need it to assign a car id.
I fix this

taras-nishchymenko commented on src/main/java/carsharing/customer/CustomerDAOImpl.java:
```diff
@@ -0,0 +1,214 @@
+    private static final String SELECT_ID_NAME_BY_CAR = "SELECT id, name FROM car WHERE company_id = ? ORDER BY id";
+    private static final String INSERT_INTO_CUSTOMER_BY_NAME = "INSERT INTO customer (name) VALUES ?";
+
+    private final DBManager DBManager;
```
Why do we need DBManager if this class extends DBManager?

ksef replied:
in new version all DAO classes doesn't extends DBManager

taras-nishchymenko commented on src/main/java/carsharing/customer/CustomerDAOImpl.java:
```diff
@@ -0,0 +1,214 @@
+        return customer;
+    }
+
+    private void setCompanyId(Customer customer, int carId) {
```
please place all private methods after public

ksef replied:
placed

taras-nishchymenko commented on src/main/java/carsharing/customer/CustomerDAOImpl.java:
```diff
@@ -0,0 +1,214 @@
+                customer = new Customer(resultSet.getString(1));
+                int carId = resultSet.getInt(2);
+                if (carId != 0) {
+                    setCompanyId(customer, carId);
```
Please avoid such shadow operations. It’s better to find company and set it here

taras-nishchymenko replied:
And consider joins instead of using several select queries

taras-nishchymenko commented on src/main/java/carsharing/customer/CustomerDAOImpl.java:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    private void setCompanyId(Customer customer, int carId) {
+        try (Connection connection = DBManager.getConnection();
```
We already have opened connection when using this method

ksef replied:
fixed

ksef committed:
1) grouped implementation and test specific dependencies;
2) deleted String BACK = "0. Back";
3) renamed pathToDb to dbName;
4) in this version all DAO classes doesn't extends DBManager
5) sorted private and public methods;
